packages/sdk/src/conversation/sso/ssoExecutionDialog.ts

```diff
@@ -0,0 +1,327 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import {
+  ComponentDialog,
+  WaterfallDialog,
+  Dialog,
+  DialogSet,
+  DialogTurnStatus,
+  DialogContext,
+} from "botbuilder-dialogs";
+import {
+  Activity,
+  ActivityTypes,
+  Channels,
+  StatePropertyAccessor,
+  Storage,
+  tokenExchangeOperationName,
+  TurnContext,
+} from "botbuilder";
+import { CommandMessage, TeamsFxBotSsoCommandHandler, TriggerPatterns } from "../interface";
+import { TeamsBotSsoPrompt, TeamsBotSsoPromptSettings } from "../../bot/teamsBotSsoPrompt";
+import { TeamsBotSsoPromptTokenResponse } from "../../bot/teamsBotSsoPromptTokenResponse";
+import { TeamsFx } from "../../core/teamsfx";
+import { v4 as uuidv4 } from "uuid";
+import { formatString } from "../../util/utils";
+import { ErrorCode, ErrorMessage, ErrorWithCode } from "../../core/errors";
+import { internalLogger } from "../../util/logger";
+
+let DIALOG_NAME = "SsoExecutionDialog";
+let TEAMS_SSO_PROMPT_ID = "TeamsFxSsoPrompt";
+let COMMAND_ROUTE_DIALOG = "CommandRouteDialog";
+
+/**
+ * Sso execution dialog, use to handle sso command
+ */
+export class SsoExecutionDialog extends ComponentDialog {
+  private dedupStorage: Storage;
+  private dedupStorageKeys: string[] = [];
+  private commandMapping: Map<string, string | RegExp | (string | RegExp)[]> = new Map<
+    string,
+    string | RegExp | (string | RegExp)[]
+  >();
+
+  /**
+   * Creates a new instance of the SsoExecutionDialog.
+   * @param dedupStorage Helper storage to remove duplicated messages
+   * @param settings The list of scopes for which the token will have access
+   * @param teamsfx {@link TeamsFx} instance for authentication
+   */
+  constructor(
+    dedupStorage: Storage,
+    ssoPromptSettings: TeamsBotSsoPromptSettings,
+    teamsfx: TeamsFx,
+    dialogName?: string
+  ) {
+    super(dialogName ?? DIALOG_NAME);
+    if (dialogName) {
+      DIALOG_NAME = dialogName;
+      TEAMS_SSO_PROMPT_ID = dialogName + TEAMS_SSO_PROMPT_ID;
+      COMMAND_ROUTE_DIALOG = dialogName + COMMAND_ROUTE_DIALOG;
+    }
+
+    this.initialDialogId = COMMAND_ROUTE_DIALOG;
+
+    this.dedupStorage = dedupStorage;
+    this.dedupStorageKeys = [];
+
+    const ssoDialog = new TeamsBotSsoPrompt(teamsfx, TEAMS_SSO_PROMPT_ID, ssoPromptSettings);
+    this.addDialog(ssoDialog);
+
+    const commandRouteDialog = new WaterfallDialog(COMMAND_ROUTE_DIALOG, [
+      this.commandRouteStep.bind(this),
+    ]);
+    this.addDialog(commandRouteDialog);
+  }
+
+  /**
+   * Add TeamsFxBotSsoCommandHandler instance
+   * @param handler TeamsFxBotSsoCommandHandler instance
+   */
+  public addCommand(handler: TeamsFxBotSsoCommandHandler): void {
+    if (!handler.commandId) {
+      handler.commandId = uuidv4();
```

Comment:
If random id always works, why expose this to user's code?
Or, what's the purpose to let user to set commandId?

Reply:
Seems typescript cannot check whether a class is type of a specific interface, so I need a property to determine whether it is a TeamsFxBotSsoCommandHandler or TeamsFxBotCommandHandler, do you have any suggestions without a property?

```
  private isSsoExecutionHandler(
    handler: TeamsFxBotCommandHandler | TeamsFxBotSsoCommandHandler
  ): boolean {
    return "commandId" in handler;
  }
```

Reply:
The potential risk is that user can add any property to the implementation. E.g., implement the `TeamsFxBotCommandHandler` interface but add one `commandId` property...

So suggest to totally separate them in middleware. Store in two arrays, or store as tuple with type, or explicitly add type to interface, etc.

Reply:
Thanks, will update